### PR TITLE
Migrate next/router to use React.createContext

### PR DESCRIPTION
**Pull request by alexandernanberg**
Fixes parts of #5716. I had some issues with the test suite but I'm fairly certain that I got it working correctly.

### CODE REVIEW

File: packages/next/package.json [hunk 1]
```diff
@@ -95,8 +95,8 @@
     "ws": "6.1.2"
   },
   "peerDependencies": {
-    "react": "^16.0.0",
-    "react-dom": "^16.0.0"
+    "react": "^16.6.0",
```
Comment (timneutkens) [inline packages/next/package.json]: 16.3 is supported in this change right?

Comment (alexandernanberg) [reply packages/next/package.json]: `contextType` (added in 16.6.0 https://reactjs.org/blog/2018/10/23/react-v-16-6.html#static-contexttype) is used in the `withRouter` hoc atm. Could refactor to use the `Consumer` component instead tho.

Comment (timneutkens) [reply packages/next/package.json]: Yeah sure please do 🙏 It makes no sense to keep the class around now that we can use a functional component 👍

**Comment by alexandernanberg** [comment]
Got any tips on how to troubleshoot if the tests are failing? The test suite keeps spawning Chrome instances 😅

**Comment by timneutkens** [comment]
You can run it locally:

```
./node_modules/next/dist/bin/next ./test/integration/basic
```

**Comment by alexandernanberg** [comment]
Ah nice thanks! Okay so I think the problem is that the `Provider` with the context isn't rendered on the server. I guess the `client/index` isn't rendered during SSR? Where would be the best place for the provider to be located?

**Comment by timneutkens** [comment]
Oh yeah I see, you didn't add it to `packages/next-server/server/render.tsx` 🕵️

**Comment by alexandernanberg** [comment]
Looks like the Azure build error is unrelated to this PR, right? 🤔

**Comment by timneutkens** [comment]
@alexandernanberg yep it fails inconsistently sometimes.

**Review by timneutkens — APPROVED** [review]

**Comment by timneutkens** [comment]
Great PR @alexandernanberg 🙏

**Comment by timneutkens** [comment]
Could you send me a DM on https://spectrum.chat/users/timneutkens

**Comment by alexandernanberg** [comment]
Done ✔️

**Comment by jaydenseric** [comment]
I'm trying to use `RouterContext` with `next@8.0.0-canary.6`, but the router is always undefined:

```jsx
import { RouterContext } from 'next/router'

const Foo = () => (
  <RouterContext.Consumer>
    {router => {
      console.log(router) // undefined
      return null
    }}
  </RouterContext.Consumer>
)
```

Can you think why?

File: packages/next/client/index.js [hunk 2]
```diff
@@ -180,7 +180,9 @@ async function doRender ({ App, Component, props, err, emitter: emitterProp = em
   // In development runtime errors are caught by react-error-overlay.
   if (process.env.NODE_ENV === 'development') {
     renderReactElement((
-      <App {...appProps} />
+      <RouterContext.Provider value={makePublicRouterInstance(router)}>
```
Comment (jaydenseric) [inline packages/next/client/index.js]: Ok, so I think this is a big breaking change for `next-graphql-react`, and probably the Apollo examples too:

https://github.com/jaydenseric/next-graphql-react/blob/v1.0.1/src/index.mjs#L92

Comment (jaydenseric) [reply packages/next/client/index.js]: Are we sure this new approach of `<RouterContext.Provider>` wrapping `<App>` is here to stay? If so, I can work it in to `next-graphql-react` and cut a new release.

**Comment by alexandernanberg** [comment]
@jaydenseric That's interesting, I can't seem to reproduce it myself. Can you provide reproducable sample repo?

File: packages/next/client/index.js (hunk 2 above)
Comment (timneutkens) [reply packages/next/client/index.js]: The way @alexandernanberg is exactly how I would have implemented it, outside of _app.js.

**Comment by jaydenseric** [comment]
@alexandernanberg in the [`with-graphql-react`](https://github.com/zeit/next.js/tree/canary/examples/with-graphql-react) example, replace the contents of `pages/index.js` with:

```jsx
import { Query } from 'graphql-react'
import { RouterContext } from 'next/router'

export default () => (
  <RouterContext.Consumer>
    {router => {
      console.log(router)
      return null
    }}
  </RouterContext.Consumer>
)
```

With `npm run dev`, in terminal you will see first `undefined` (bad) for the GraphQL preload and then a router instance when the actual SSR happens:

screen shot 2019-01-13 at 9 50 03 am

I imagine this reproduction would work for the [`with-apollo`](https://github.com/zeit/next.js/tree/canary/examples/with-apollo) examples too:

https://github.com/zeit/next.js/blob/v8.0.0-canary.6/examples/with-apollo/lib/with-apollo-client.js#L27

**Pull request merged**